# Preprocesamiento de datos

Se importan las librerias

In [ ]:
import pandas as pd
import numpy as np

Se monta el google drive en el colab

In [ ]:
from google.colab import drive 

drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Se inserta la dirección del archivo adult.csv y se mando a llamar su encabezado.

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Artek/5toTrimestre/Reconocimiento de Patrones/Semana 6/Recursos de clase/adult.csv', encoding='latin-1')
df.head()

age  workclass  fnlwgt     education  educational-num      marital-status  \
0   25    Private  226802          11th                7       Never-married   
1   38    Private   89814       HS-grad                9  Married-civ-spouse   
2   28  Local-gov  336951    Assoc-acdm               12  Married-civ-spouse   
3   44    Private  160323  Some-college               10  Married-civ-spouse   
4   18          ?  103497  Some-college               10       Never-married   

          occupation relationship   race  gender  capital-gain  capital-loss  \
0  Machine-op-inspct    Own-child  Black    Male             0             0   
1    Farming-fishing      Husband  White    Male             0             0   
2    Protective-serv      Husband  White    Male             0             0   
3  Machine-op-inspct      Husband  Black    Male          7688             0   
4                  ?    Own-child  White  Female             0             0   

   hours-per-week native-country income  
0              40  United-States  <=50K  
1              50  United-States  <=50K  
2              40  United-States   >50K  
3              40  United-States   >50K  
4              30  United-States  <=50K

Se manda a llamar shape para concoer el número de elementos que contiene el conjunto de datos.

In [ ]:
df.shape

(48842, 15)

En este caso, el conjunto esta conformado por 48842 filas y 15 columnas.

Como podemos observar en ciertas partes aparece el símbolo ?. Por lo que podemos pensar que es NaN.

In [ ]:
df = df.replace('?', np.nan)
df.head()

age  workclass  fnlwgt     education  educational-num      marital-status  \
0   25    Private  226802          11th                7       Never-married   
1   38    Private   89814       HS-grad                9  Married-civ-spouse   
2   28  Local-gov  336951    Assoc-acdm               12  Married-civ-spouse   
3   44    Private  160323  Some-college               10  Married-civ-spouse   
4   18        NaN  103497  Some-college               10       Never-married   

          occupation relationship   race  gender  capital-gain  capital-loss  \
0  Machine-op-inspct    Own-child  Black    Male             0             0   
1    Farming-fishing      Husband  White    Male             0             0   
2    Protective-serv      Husband  White    Male             0             0   
3  Machine-op-inspct      Husband  Black    Male          7688             0   
4                NaN    Own-child  White  Female             0             0   

   hours-per-week native-country income  
0              40  United-States  <=50K  
1              50  United-States  <=50K  
2              40  United-States   >50K  
3              40  United-States   >50K  
4              30  United-States  <=50K

Para observar la cantidad de datos nulos

In [ ]:
nan_per_col = df.isna().sum()

In [ ]:
print(nan_per_col)

age                   0
workclass          2799
fnlwgt                0
education             0
educational-num       0
marital-status        0
occupation         2809
relationship          0
race                  0
gender                0
capital-gain          0
capital-loss          0
hours-per-week        0
native-country      857
income                0
dtype: int64


## Para reemplazar datos nulos existen las siguientes opciones:

Primero se deben de convertir los valores categoricos en numéricos. 

In [ ]:
mappings = {
    'workclass': {val: i for i, val in enumerate(df['workclass'].dropna().unique(), start=1)},
    'education': {val: i for i, val in enumerate(df['education'].dropna().unique(), start=1)},
    'marital-status': {val: i for i, val in enumerate(df['marital-status'].dropna().unique(), start=1)},
    'occupation': {val: i for i, val in enumerate(df['occupation'].dropna().unique(), start=1)},
    'relationship': {val: i for i, val in enumerate(df['relationship'].dropna().unique(), start=1)},
    'race': {val: i for i, val in enumerate(df['race'].dropna().unique(), start=1)},
    'native-country': {val: i for i, val in enumerate(df['native-country'].dropna().unique(), start=1)},
    'gender': {val: i for i, val in enumerate(df['gender'].dropna().unique(), start=1)}
}


for column in mappings:
    i = 1
    for val in df[column].dropna().unique():
        mappings[column][val] = i
        i += 1


In [ ]:
for column in df1.columns:
    if column in mappings:
        unique = df1[column].dropna().unique()
        dictionary = {value: key for key, value in enumerate(unique, start=1)}
        df1[column] = df1[column].map(dictionary)
        mappings[column] = dictionary

df1.head()


age  workclass  fnlwgt  education  educational-num  marital-status  \
0   25          1  226802          1                7               1   
1   38          1   89814          2                9               2   
2   28          2  336951          3               12               2   
3   44          1  160323          4               10               2   
4   18          3  103497          4               10               1   

   occupation  relationship  race  gender  capital-gain  capital-loss  \
0           1             1     1       1             0             0   
1           2             2     2       1             0             0   
2           3             2     2       1             0             0   
3           1             2     1       1          7688             0   
4           4             1     2       2             0             0   

   hours-per-week  native-country income  
0              40               1  <=50K  
1              50               1  <=50K  
2              40               1   >50K  
3              40               1   >50K  
4              30               1  <=50K

#Opcion 1 rellenar datos con media aritmentica

In [ ]:
# Reemplaza los valores NaN con la media aritmética de la columna
df1 = df.fillna(df.mean())
df1.head()

<ipython-input-8-c8312ed0acec>:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df1 = df.fillna(df.mean())


age  workclass  fnlwgt     education  educational-num      marital-status  \
0   25    Private  226802          11th                7       Never-married   
1   38    Private   89814       HS-grad                9  Married-civ-spouse   
2   28  Local-gov  336951    Assoc-acdm               12  Married-civ-spouse   
3   44    Private  160323  Some-college               10  Married-civ-spouse   
4   18        NaN  103497  Some-college               10       Never-married   

          occupation relationship   race  gender  capital-gain  capital-loss  \
0  Machine-op-inspct    Own-child  Black    Male             0             0   
1    Farming-fishing      Husband  White    Male             0             0   
2    Protective-serv      Husband  White    Male             0             0   
3  Machine-op-inspct      Husband  Black    Male          7688             0   
4                NaN    Own-child  White  Female             0             0   

   hours-per-week native-country income  
0              40  United-States  <=50K  
1              50  United-States  <=50K  
2              40  United-States   >50K  
3              40  United-States   >50K  
4              30  United-States  <=50K

In [ ]:
df1 = df1.fillna(df1.mean())

df1.head()

<ipython-input-25-e95f093f78ca>:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df1 = df1.fillna(df1.mean())


age  workclass  fnlwgt  education  educational-num  marital-status  \
0   25   1.000000  226802          1                7               1   
1   38   1.000000   89814          2                9               2   
2   28   2.000000  336951          3               12               2   
3   44   1.000000  160323          4               10               2   
4   18   1.689595  103497          4               10               1   

   occupation  relationship  race  gender  capital-gain  capital-loss  \
0    1.000000             1     1       1             0             0   
1    2.000000             2     2       1             0             0   
2    3.000000             2     2       1             0             0   
3    1.000000             2     1       1          7688             0   
4    6.739991             1     2       2             0             0   

   hours-per-week  native-country income  
0              40             1.0  <=50K  
1              50             1.0  <=50K  
2              40             1.0   >50K  
3              40             1.0   >50K  
4              30             1.0  <=50K

In [ ]:
nulos_por_columna = df1.isna().sum()
print(nulos_por_columna)

age                0
workclass          0
fnlwgt             0
education          0
educational-num    0
marital-status     0
occupation         0
relationship       0
race               0
gender             0
capital-gain       0
capital-loss       0
hours-per-week     0
native-country     0
income             0
dtype: int64


# Opcion 2 Remplazar los Nan con los valores de Mediana

In [ ]:
# Reemplaza los valores NaN con la mediana de la columna
df2 = df1.fillna(df1.median())
df2.head()

<ipython-input-27-af455acd13ad>:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df2 = df1.fillna(df1.median())


age  workclass  fnlwgt  education  educational-num  marital-status  \
0   25   1.000000  226802          1                7               1   
1   38   1.000000   89814          2                9               2   
2   28   2.000000  336951          3               12               2   
3   44   1.000000  160323          4               10               2   
4   18   1.689595  103497          4               10               1   

   occupation  relationship  race  gender  capital-gain  capital-loss  \
0    1.000000             1     1       1             0             0   
1    2.000000             2     2       1             0             0   
2    3.000000             2     2       1             0             0   
3    1.000000             2     1       1          7688             0   
4    6.739991             1     2       2             0             0   

   hours-per-week  native-country income  
0              40             1.0  <=50K  
1              50             1.0  <=50K  
2              40             1.0   >50K  
3              40             1.0   >50K  
4              30             1.0  <=50K

In [ ]:
nulos_por_columna = df2.isna().sum()
print(nulos_por_columna)

age                0
workclass          0
fnlwgt             0
education          0
educational-num    0
marital-status     0
occupation         0
relationship       0
race               0
gender             0
capital-gain       0
capital-loss       0
hours-per-week     0
native-country     0
income             0
dtype: int64


#Opcion 3 Remplazar los Nan con los valores mas frecuentes

In [ ]:
# Reemplaza los valores NaN con el valor más frecuente de la columna
df3 = df.fillna(df.mode().iloc[0])
df3.head()

age  workclass  fnlwgt     education  educational-num      marital-status  \
0   25    Private  226802          11th                7       Never-married   
1   38    Private   89814       HS-grad                9  Married-civ-spouse   
2   28  Local-gov  336951    Assoc-acdm               12  Married-civ-spouse   
3   44    Private  160323  Some-college               10  Married-civ-spouse   
4   18    Private  103497  Some-college               10       Never-married   

          occupation relationship   race  gender  capital-gain  capital-loss  \
0  Machine-op-inspct    Own-child  Black    Male             0             0   
1    Farming-fishing      Husband  White    Male             0             0   
2    Protective-serv      Husband  White    Male             0             0   
3  Machine-op-inspct      Husband  Black    Male          7688             0   
4     Prof-specialty    Own-child  White  Female             0             0   

   hours-per-week native-country income  
0              40  United-States  <=50K  
1              50  United-States  <=50K  
2              40  United-States   >50K  
3              40  United-States   >50K  
4              30  United-States  <=50K

In [ ]:
nulos_por_columna = df3.isna().sum()
print(nulos_por_columna)

age                0
workclass          0
fnlwgt             0
education          0
educational-num    0
marital-status     0
occupation         0
relationship       0
race               0
gender             0
capital-gain       0
capital-loss       0
hours-per-week     0
native-country     0
income             0
dtype: int64


#Eliminar Filas con valores Nulos

In [ ]:
df4 = df.dropna()
df4.head()

age  workclass  fnlwgt     education  educational-num      marital-status  \
0   25    Private  226802          11th                7       Never-married   
1   38    Private   89814       HS-grad                9  Married-civ-spouse   
2   28  Local-gov  336951    Assoc-acdm               12  Married-civ-spouse   
3   44    Private  160323  Some-college               10  Married-civ-spouse   
5   34    Private  198693          10th                6       Never-married   

          occupation   relationship   race gender  capital-gain  capital-loss  \
0  Machine-op-inspct      Own-child  Black   Male             0             0   
1    Farming-fishing        Husband  White   Male             0             0   
2    Protective-serv        Husband  White   Male             0             0   
3  Machine-op-inspct        Husband  Black   Male          7688             0   
5      Other-service  Not-in-family  White   Male             0             0   

   hours-per-week native-country income  
0              40  United-States  <=50K  
1              50  United-States  <=50K  
2              40  United-States   >50K  
3              40  United-States   >50K  
5              30  United-States  <=50K

In [ ]:
df4.shape

(45222, 15)

In [ ]:
nulos_por_columna = df4.isna().sum()
print(nulos_por_columna)

age                0
workclass          0
fnlwgt             0
education          0
educational-num    0
marital-status     0
occupation         0
relationship       0
race               0
gender             0
capital-gain       0
capital-loss       0
hours-per-week     0
native-country     0
income             0
dtype: int64


In [ ]:
# mostrar las filas originales con valores NaN
filas_nulas = df.loc[df.isnull().any(axis=1)]
print("Filas con valores nulos:\n", filas_nulas)

# reemplazar valores NaN con la media
df_con_media = df.fillna(df.mean())

# mostrar las filas eliminadas debido a NaN
filas_eliminadas = df_con_media.dropna(how='any')
print("\nFilas eliminadas con NaN:\n", filas_eliminadas)

# reemplazar valores NaN con la mediana
df_con_mediana = df.fillna(df.median())

# mostrar las filas eliminadas debido a NaN
filas_eliminadas = df_con_mediana.dropna(how='any')
print("\nFilas eliminadas con NaN:\n", filas_eliminadas)

# reemplazar valores NaN con el valor más común
df_con_frecuentes = df.fillna(df.mode().iloc[0])

# mostrar las filas eliminadas debido a NaN
filas_eliminadas = df_con_frecuentes.dropna(how='any')
print("\nFilas eliminadas con NaN:\n", filas_eliminadas)


Filas con valores nulos:
        age workclass  fnlwgt     education  educational-num  \
4       18       NaN  103497  Some-college               10   
6       29       NaN  227026       HS-grad                9   
13      58       NaN  299831       HS-grad                9   
19      40   Private   85019     Doctorate               16   
22      72       NaN  132015       7th-8th                4   
...    ...       ...     ...           ...              ...   
48811   35       NaN  320084     Bachelors               13   
48812   30       NaN   33811     Bachelors               13   
48820   71       NaN  287372     Doctorate               16   
48822   41       NaN  202822       HS-grad                9   
48823   72       NaN  129912       HS-grad                9   

           marital-status      occupation   relationship                race  \
4           Never-married             NaN      Own-child               White   
6           Never-married             NaN      Unmarried 

<ipython-input-36-89e22678ee64>:6: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df_con_media = df.fillna(df.mean())
<ipython-input-36-89e22678ee64>:13: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df_con_mediana = df.fillna(df.median())



Filas eliminadas con NaN:
        age     workclass  fnlwgt     education  educational-num  \
0       25       Private  226802          11th                7   
1       38       Private   89814       HS-grad                9   
2       28     Local-gov  336951    Assoc-acdm               12   
3       44       Private  160323  Some-college               10   
5       34       Private  198693          10th                6   
...    ...           ...     ...           ...              ...   
48837   27       Private  257302    Assoc-acdm               12   
48838   40       Private  154374       HS-grad                9   
48839   58       Private  151910       HS-grad                9   
48840   22       Private  201490       HS-grad                9   
48841   52  Self-emp-inc  287927       HS-grad                9   

           marital-status         occupation   relationship   race  gender  \
0           Never-married  Machine-op-inspct      Own-child  Black    Male   
1      Marr